docs/source/cuda/cuda_ffi.rst

```diff
@@ -0,0 +1,148 @@
+PTX, and binary objects (cubins, fat binaries, etc.) are directly supported;
+sources in other languages must be compiled to PTX first. The constituent parts
+of a Python kernel call to a foreign device function are:
+
+- The device function implementation in a foreign language (e.g. CUDA C).
+- A declaration of the device function in Python.
+- A kernel that links with and calls the foreign function.
+
+
+Device function ABI
+-------------------
+
+Numba's ABI for calling device functions defines the following prototype in
+C/C++:
+
+.. code:: C
+
+   extern "C"
+   __device__ int
+   function(
+     T* return_value,
+     ...
+   );
+
+
+Components of the prototype are as follows:
+
+- ``extern "C"`` is used to prevent name-mangling so that it is easy to declare
+  the function in Python. It can be removed, but then the mangled name must be
+  used in the declaration of the function in Python.
+- ``__device__`` is required to define the function as a device function.
+- The return value is always of type ``int``, and is used to signal whether a
+  Python exception occurred. Since Python exceptions don't occur in foreign
+  functions, this should always be set to 0 by the callee.
+- The first argument is a pointer to the return value of type ``T``, which is
+  passed in by the caller. If the function returns a value, the pointee should
+  be set by the callee to store the return value.
+- Subsequent arguments should match the types and order of arguments passed to
+  the function from the Python kernel.
+
+Functions written in other languages must compile to PTX that conforms to this
+prototype specification.
+
+A function that accepts two floats and returns a float would have the following
+prototype:
+
+.. code:: C
+
+   extern "C"
+   __device__ int
+   mul_f32_f32(
+     float* return_value,
+     float x,
+     float y
+   )
+
+
+Declaration in Python
+---------------------
+
+To declare a foreign device function in Python, use :func:`declare_device()
+<numba.cuda.declare_device>`:
+
+.. autofunction:: numba.cuda.declare_device
+
+The returned descriptor name need not match the name of the foreign function.
+For example, when:
+
+.. code::
+
+   mul = cuda.declare_device('mul_f32_f32', 'float32(float32, float32)')
+
+is declared, calling ``mul(a, b)`` inside a kernel will translate into a call to
+``mul_f32_f32(a, b)`` in the compiled code.
+
+
+Linking and Calling functions
+-----------------------------
+
+The ``link`` keyword argument of the :func:`@cuda.jit <numba.cuda.jit>`
+decorator accepts a list of files; files whose name ends in ``.cu`` will be
+compiled with the `NVIDIA Runtime Compiler (NVRTC)
+<https://docs.nvidia.com/cuda/nvrtc/index.html>`_ and linked into the kernel as
+PTX; other files will be passed directly to the CUDA Linker.
+
+For example, the following kernel calls the ``mul()`` function declared above
+with the implementation ``mul_f32_f32()`` in a file called ``functions.cu``:
```

Comment:
Does something need to be said about:
* The required presence of `nvrtc` on the path (and where to get it).
* How specified files are resolved via paths, am guessing working directory only or full path required?

Reply:
I've now added a note about the relative / absolute path specification for linked files, and also about the need for NVRTC. I didn't get too specific about NVRTC because I felt it would get needlessly complex and confusing trying to be specific about the possibilities across platforms and versions.

Reply:
Thanks for adding the path specification. RE the `nvrtc` binary location, this seems reasonable, it's quite an advanced use case and it needing to be on the path would likely be an implicit requirement, how this is resolved appropriately is understandably user specific.